testnet/stacks-node/src/tests/nakamoto_integrations.rs

```diff
@@ -8030,3 +8030,134 @@ fn utxo_check_on_startup_recover() {
+        naka_submitted_commits: commits_submitted,
+        naka_proposed_blocks: proposals_submitted,
+        ..
+    } = run_loop.counters();
+
+    let coord_channel = run_loop.coordinator_channels();
+
+    let run_loop_thread = thread::spawn(move || run_loop.start(None, 0));
+    let mut signers = TestSigners::new(vec![signer_sk.clone()]);
+    wait_for_runloop(&blocks_processed);
+    boot_to_epoch_3(
+        &conf,
+        &blocks_processed,
+        &[stacker_sk],
+        &[signer_sk],
+        &mut Some(&mut signers),
+        &mut btc_regtest_controller,
+    );
+
+    info!("------------------------- Reached Epoch 3.0 -------------------------");
+
+    blind_signer(&conf, &signers, proposals_submitted);
+    wait_for_first_naka_block_commit(60, &commits_submitted);
+
+    // TODO (hack) instantiate the sortdb in the burnchain
+    _ = btc_regtest_controller.sortdb_mut();
+
+    info!("------------------------- Setup finished, run test -------------------------");
+
+    let naka_tenures = 20;
+    let pre_naka_reward_cycle = 1;
+    let http_origin = format!("http://{}", &conf.node.rpc_bind);
+
+    let get_v3_signer = |pubkey: &Secp256k1PublicKey, reward_cycle: u64| {
+        let url = &format!(
+            "{http_origin}/v3/signer/{pk}/{reward_cycle}",
+            pk = pubkey.to_hex()
+        );
+        info!("Send request: GET {url}");
+        reqwest::blocking::get(url)
+            .unwrap_or_else(|e| panic!("GET request failed: {e}"))
+            .text()
+            .inspect(|response| info!("Recieved response: GET {url} -> {response}"))
+            .expect("Empty response")
+            .parse::<u64>()
+            .unwrap_or_else(|e| panic!("Failed to parse response as `u64`: {e}"))
+    };
+
+    // Check reward cycle 1, should be 0 (pre-nakamoto)
+    let blocks_signed_pre_naka = get_v3_signer(&signer_pubkey, pre_naka_reward_cycle);
+    assert_eq!(blocks_signed_pre_naka, 0);
+
+    // Keep track of reward cycles encountered
+    let mut reward_cycles = HashSet::new();
+
+    // Mine some nakamoto tenures
+    for _ in 0..naka_tenures {
+        next_block_and_mine_commit(
+            &mut btc_regtest_controller,
+            60,
+            &coord_channel,
+            &commits_submitted,
+        )
+        .unwrap();
+        let block_height = btc_regtest_controller.get_headers_height();
+        let reward_cycle = btc_regtest_controller
+            .get_burnchain()
+            .block_height_to_reward_cycle(block_height)
+            .unwrap();
+        reward_cycles.insert(reward_cycle);
+    }
+
+    // Make sure we got a couple cycles
+    assert!(reward_cycles.len() > 1);
+    assert!(!reward_cycles.contains(&pre_naka_reward_cycle));
+
+    // Since we have only one signer, it must be signing at least 1 block per reward cycle
+    for reward_cycle in reward_cycles.into_iter() {
+        let blocks_signed = get_v3_signer(&signer_pubkey, reward_cycle);
+        assert_ne!(blocks_signed, 0);
```

Comment:
Is there a way to assert the exact number of blocks it signed?

Reply:
Yep. Done.